crypto/fipsmodule/rand/entropy_pool_test.cc

```diff
@@ -0,0 +1,190 @@
+  EXPECT_EQ(entropy_pool.index_read, (size_t) ENTROPY_POOL_SIZE);
+  EXPECT_EQ(OPENSSL_memcmp(entropy_pool.pool, zero_entropy_pool, ENTROPY_POOL_SIZE), 0);
+  EXPECT_EQ(OPENSSL_memcmp(entropy_buffer, fake_entropy, ENTROPY_POOL_SIZE), 0);
+
+  RAND_entropy_pool_zeroize(&entropy_pool);
+  EXPECT_EQ(entropy_pool.capacity, (size_t) 0);
+  EXPECT_EQ(entropy_pool.valid_available, (size_t) 0);
+  EXPECT_EQ(entropy_pool.index_read, (size_t) 0);
+  EXPECT_EQ(OPENSSL_memcmp(entropy_pool.pool, zero_entropy_pool, ENTROPY_POOL_SIZE), 0);
+
+  // Initialize again but now consume a total of |ENTROPY_POOL_SIZE| bytes of
+  // entropy over two get invocations.
+  RAND_entropy_pool_init(&entropy_pool);
+  RAND_entropy_pool_add(&entropy_pool, fake_entropy);
+
+  OPENSSL_cleanse(entropy_buffer, ENTROPY_POOL_SIZE);
+  EXPECT_TRUE(RAND_entropy_pool_get(&entropy_pool, entropy_buffer, entropy_pool_size_halfed));
+  EXPECT_EQ(entropy_pool.capacity, (size_t) ENTROPY_POOL_SIZE);
+  EXPECT_EQ(entropy_pool.valid_available, (size_t) entropy_pool_size_halfed);
+  EXPECT_EQ(entropy_pool.index_read, (size_t) entropy_pool_size_halfed);
+  // Should have zeroized first half of pool. Second half should still contain
+  // the added entropy
+  EXPECT_EQ(OPENSSL_memcmp(entropy_pool.pool, zero_entropy_pool, entropy_pool_size_halfed), 0);
+  EXPECT_EQ(OPENSSL_memcmp(entropy_pool.pool + entropy_pool_size_halfed,
+    fake_entropy + entropy_pool_size_halfed, entropy_pool_size_halfed), 0);
+  // The reverse should be true for the output buffer.
+  EXPECT_EQ(OPENSSL_memcmp(entropy_buffer,
+    fake_entropy, entropy_pool_size_halfed), 0);
+  EXPECT_EQ(OPENSSL_memcmp(entropy_buffer + entropy_pool_size_halfed,
+    zero_entropy_pool, entropy_pool_size_halfed), 0);
+
+  EXPECT_TRUE(RAND_entropy_pool_get(&entropy_pool,
+    entropy_buffer + entropy_pool_size_halfed, entropy_pool_size_halfed));
+  EXPECT_EQ(entropy_pool.capacity, (size_t) ENTROPY_POOL_SIZE);
+  EXPECT_EQ(entropy_pool.valid_available, (size_t) 0);
+  EXPECT_EQ(entropy_pool.index_read, (size_t) ENTROPY_POOL_SIZE);
+  // Entire pool must now be zeroized and output buffer full of entropy.
+  EXPECT_EQ(OPENSSL_memcmp(entropy_pool.pool, zero_entropy_pool, ENTROPY_POOL_SIZE), 0);
+  EXPECT_EQ(OPENSSL_memcmp(entropy_buffer, fake_entropy, ENTROPY_POOL_SIZE), 0);
+}
+
+TEST(EntropyPool, BasicFailure) {
+  uint8_t entropy_buffer[ENTROPY_POOL_SIZE] = {0};
+  struct entropy_pool entropy_pool;
+  uint8_t fake_entropy[ENTROPY_POOL_SIZE] = {0};
+  fill_fake_entropy(fake_entropy, 'A');
+
+  RAND_entropy_pool_init(&entropy_pool);
+  RAND_entropy_pool_add(&entropy_pool, fake_entropy);
+
+  // Input validation. Cannot:
+  //  * Consume strictly more than |ENTROPY_POOL_SIZE| bytes per invocation.
+  //  * Consume zero bytes.
+  //  * Use NULL as argument for entropy pool.
+  //  * Use NULL as argument for output buffer.
+  EXPECT_FALSE(RAND_entropy_pool_get(&entropy_pool, entropy_buffer, ENTROPY_POOL_SIZE+1));
```

Comment:
Is it worth checking entropy_pool wasn't touched and still has the original values?

Reply:
OK